returnn/torch/data/pipeline.py

```diff
@@ -309,6 +311,105 @@ def __iter__(self):
+        :param length_key: data key to take as length measure
+        """
+        self._dataset = dataset
+        self._length_key = length_key
+
+        assert buckets, "empty bucket batching configuration"
+        if not all(size > 0 and max_seqs > 0 for size, max_seqs in buckets):
+            raise ValueError(f"bucket sizes and max seqs in bucket must be positive")
+        self._max_seq_lens, self._max_bucket_sizes = zip(*sorted(buckets, key=lambda b: b[0]))
+        assert len(set(self._max_seq_lens)) == len(self._max_seq_lens), "seq len boundaries must all be unique"
+
+    def __iter__(self):
+        """:return: generator applying bucket ordering on the data"""
+        buckets: List[List[Dict[str, numpy.ndarray]]] = [[] for _ in range(len(self._max_seq_lens))]
+
+        for data_dict in self._dataset:
+            data_dict: Dict[str, numpy.ndarray]
+            data_size = data_dict[self._length_key].shape[0]
+            bucket_idx = bisect.bisect_left(self._max_seq_lens, data_size)
+            if bucket_idx >= len(self._max_seq_lens):
+                # seg is too long, drop it
+                continue
+            buckets[bucket_idx].append(data_dict)
+            if len(buckets[bucket_idx]) >= self._max_bucket_sizes[bucket_idx]:
+                yield buckets[bucket_idx]
+                buckets[bucket_idx] = []
+
+        non_empty_buckets = [b for b in buckets if b]
+        yield from non_empty_buckets
+
+
+def get_batching_iterable_dataset_from_config(
+    *, dataset: torch.utils.data.IterableDataset, config: Config, train: bool
+) -> torch.utils.data.IterableDataset:
+    """
+    Batches the segments in the given dataset given the config settings.
+
+    :param dataset: dataset whose segments to group into batches
+    :param config: RETURNN config
+    :param train: whether we are in training or in inference mode
+    """
+    custom_batching = config.typed_value("custom_batching", None)
+    if custom_batching is None:
+        batch_size = config.typed_value("batch_size", -1)
+        batch_size = config.typed_value(f"batch_size_{'train' if train else 'dev'}", batch_size)
+        assert batch_size != -1, f"batch_size or batch_size_{'train' if train else 'dev'} not defined in config"
+        max_seqs = config.typed_value("max_seqs", -1)
+        batches_dataset = BatchingIterDataPipe(dataset, batch_size=batch_size, max_seqs=max_seqs)
+        return batches_dataset
+
+    if isinstance(custom_batching, dict):
+        assert "class" in custom_batching
+        batching_args = custom_batching.copy()
+        type_or_name = batching_args.pop("class")
+        if isinstance(type_or_name, str):
+            cls = globals()[type_or_name]
+        elif isinstance(type_or_name, (type, Callable)):
+            # custom types need to be forward compatible
+            batching_args.update(get_fwd_compat_kwargs())
+            batching_args["train"] = train
```

Comment:
Remove this. If the user wants the custom callable, it can be used directly without a dict.
```suggestion
```
Even having support for `type` here is maybe also not really necessary, not sure.